llrt_core/src/custom_resolver.rs

```diff
@@ -0,0 +1,570 @@
+    package_json_path.push_str(scope);
+    package_json_path.push_str("/package.json");
+
+    let (scope, name) = if name != "." && !Path::new(&package_json_path).exists() {
+        package_json_path.truncate(base_path_length);
+        package_json_path.push_str(x);
+        package_json_path.push_str("/package.json");
+        (x, ".")
+    } else {
+        (scope, name.as_str())
+    };
+
+    if !Path::new(&package_json_path).exists() {
+        return Err(Error::new_resolving(dir.to_string(), x.to_string()));
+    };
+
+    //3. Parse DIR/NAME/package.json, and look for "exports" field.
+    //4. If "exports" is null or undefined, return.
+    //5. let MATCH = PACKAGE_EXPORTS_RESOLVE(pathToFileURL(DIR/NAME), "." + SUBPATH,
+    //   `package.json` "exports", ["node", "require"]) <a href="esm.md#resolver-algorithm-specification">defined in the ESM resolver</a>.
+    //6. RESOLVE_ESM_MATCH(MATCH)
+    let mut package_json = fs::read(&package_json_path).unwrap_or_default();
+    let package_json = simd_json::to_borrowed_value(&mut package_json).or_throw(ctx)?;
+
+    let module_path = package_exports_resolve(&package_json, name, is_esm)?;
+
+    Ok([dir, "/", scope, "/", module_path].concat())
+}
+
+// LOAD_PACKAGE_SELF(X, DIR)
+fn load_package_self(ctx: &Ctx<'_>, x: &str, dir: &str, is_esm: bool) -> Result<Option<String>> {
+    trace!("|  load_package_self(x, dir): ({}, {})", x, dir);
+
+    let (scope, name) = match x.split_once('/') {
+        Some((s, n)) => (s, ["./", n].concat()),
+        None => (x, ".".to_string()),
+    };
+    let name = name.as_str();
+
+    // 1. Find the closest package scope SCOPE to DIR.
+    let package_json_path = match find_the_closest_package_scope(dir) {
+        // 2. If no scope was found, return.
+        None => {
+            return Ok(None);
+        },
+        Some(path) => {
+            // 3. If the SCOPE/package.json "exports" is null or undefined, return.
+            if let Ok(None) = get_exports_field(ctx, &path) {
+                return Ok(None);
+            }
+            // 4. If the SCOPE/package.json "name" is not the first segment of X, return.
+            if let Some(val) = get_main_field(ctx, &path)? {
+                if val != scope {
+                    return Ok(None);
+                }
+            }
+            path
+        },
+    };
+    // 5. let MATCH = PACKAGE_EXPORTS_RESOLVE(pathToFileURL(SCOPE),
+    //    "." + X.slice("name".length), `package.json` "exports", ["node", "require"])
+    //    <a href="esm.md#resolver-algorithm-specification">defined in the ESM resolver</a>.
+    // 6. RESOLVE_ESM_MATCH(MATCH)
+    let mut package_json = fs::read(&package_json_path).unwrap_or_default();
+    let package_json = simd_json::to_borrowed_value(&mut package_json).or_throw(ctx)?;
+
+    if let Ok(path) = package_exports_resolve(&package_json, name, is_esm) {
+        trace!("|  load_package_self(2.c): {}", path);
+        return Ok(Some(path.to_string()));
+    }
```

Comment:
Avoid parsing and loading the same json file 3 times. Parse json once as simd::borrowed and pass to get_main_field, get_exports_field etc

Reply:
We have responded to your request.